### ISSUE TITLE
Perf: OmniSharp takes several minutes to load projects 

### ISSUE THREAD

**Issue by natemcmaster** [OmniSharp/omnisharp-roslyn#926]
OmniSharp is timing out on loading solutions with large files within the default timeout in VS code (60 seconds). The timeout can be increased, but the solution still takes several minutes to load. On some solutions, it takes 3-4 minutes to load the project system.

**Repro:**
git clone https://github.com/aspnet/Blazor-Hackathon
dotnet restore Blazor.Core.sln
Open VS Code and try to load *Blazor.Core.sln*

Canonical link: https://github.com/aspnet/Blazor-Hackathon/tree/a8c341fcb5a519fb7c931eaa9ba23bde560433ef

**Result**
The output log shows OmniSharp trying to load the project correctly, but it times out after 60 seconds with an error printed to the Output > Omnisharp Log window in in VS Code.

**Workaround**
Increase project load timeout to 4 minutes
```json
{
  "omnisharp.projectLoadTimeout": 240
}
```

**Details**
VS Code: 1.14.2
C# Extension: 1.12.0-beta3
Mono: 5.2.0.196
OS: macOS 10.12.5
dotnet: 2.0.0-preview2-06497

**Comment by davidfowl** [comment]
I'm hitting this too. Larger solutions need more time to load.

**Comment by Figbash** [comment]
Is there anything that can be done to speed this up locally? It's taking around 2.5 minutes on my project.

**Comment by DustinCampbell** [comment]
This is going to get faster soon. We've just taken newer MSBuild bits, which have some performance enhancements. In addition, you can install the latest Mono MDK and load OmniSharp via that Mono rather than the stripped-down Mono that OmniSharp includes to get a speed boost. If you run OmniSharp on an official Mono install, it will run on native AOT-compiled images rather than JIT.

**Comment by Figbash** [comment]
OK, using official Mono shaves off ~10 seconds for me, I am looking forward to the MSBuild enhancements!
For me at least 35 seconds of it is taken between:
[info]: OmniSharp.Cake.CakeProjectSystem
        Detecting Cake files in '*****'.
[info]: OmniSharp.Cake.CakeProjectSystem
        Could not find any Cake files
Is there possibly a way to disable this detection?

**Comment by DustinCampbell** [comment]
Wow! That's pretty significant! I don't know if there's any way to disable it. @mholo65: Do you have any ideas?

**Comment by filipw** [comment]
you can always disable any project system (msbuild, dotnet, script, cake) using the following syntax in the [omnisharp.json](https://github.com/OmniSharp/omnisharp-roslyn/wiki/Configuration-Options#global-omnisharpjson) file.

```
{
   "{project key}": {  
      "enabled": false
   }
}
```

So in this case it would be:

```
{
   "cake": {  
      "enabled": false
   }
}
```

Then to project system wouldn't even get initialized.

**Comment by mholo65** [comment]
@DustinCampbell @Figbash yeah problem is that both [csx project system](https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Script/ScriptProjectSystem.cs#L87) and [cake project system](https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Cake/CakeProjectSystem.cs#L75) searches all files in all directories for `csx` or `cake` files. This is a known issue and reported [here](https://github.com/OmniSharp/omnisharp-roslyn/issues/896). 

ping @filipw

**Comment by Figbash** [comment]
@filipw Thanks, that helped shave off 35 seconds.
@DustinCampbell @mholo65 Yes that is definitely the issue, now it's OmniSharp.DotNet.DotNetProjectSystem taking the time I believe, but I assume I need that one.
My solution files are all in the same root folder as my workspace, so searching through all the subfolders to find them is fruitless, if that is what it is doing. Perhaps an option to only search 1 level, or to respect VSCode's files.exclude option in the VSCode plugin?
It is a Unity project and has ~850K(!) files in the various subfolders.
Thanks for looking into it!

**Comment by DustinCampbell** [comment]
@Figbash: You only need the DotNetProejctSystem if you are using legacy project.json projects. If you're doing Unity work, you can safely disable this one too.

**Comment by savpek** [comment]
I tried to write prototype that loads projects concurrently. Expensive/slow part is project instance build step that takes majority of time during load. However it seems that build cannot be executed concurrently `The operation cannot be completed because a build is already in progress` if i try execute `var buildResult = projectInstance.Build(...)` in parallel (`MSB.Execution.ProjectInstance`).

Is there way to build concurrently in msbuild workspace at all? Steps like updating projects takes only few percent of total time during startup, so this build is the part of optimize if any 🤔

**Comment by vchirikov** [comment]
We have about 260 projects, and omnisharp is completely unusable, because of slow loading of projects.
I know that msbuild is slow, and it can takeup to 30 sec to evaluate of all csproj properties (with 1 thread) but with omnisharp `ProjectManager` init took 15 min.

**Comment by TCROC** [comment]
Is there an update on this?  It takes quite a while for our projects to load with Omnisharp

**Comment by TCROC** [comment]
Unity Development

**Comment by TCROC** [comment]
Just checking in to see if there are any updates on this issue.  It is definitely something our team struggles with still.

**Comment by Apppppppple** [comment]
Yep - Even just opening a file can take several minutes for omnisharp to load, if it loads at all

**Comment by TCROC** [comment]
And before this is written off as a hardware issue on the user's end, here are my hardware specs:

Intel(R) Core(TM) i7-9700K CPU @ 3.60GHz
32 GB RAM
NVME SSD
GPU RTX 2070 Super

So my hardware should be able to load up my projects faster than it does.  Other IDEs that I use for the same project load up much faster.

**Comment by jparadnikas** [comment]
Waiting like 30secs for 30k lines

**Comment by Thiago-Morais** [comment]
Is there any update on this?

**Comment by loeffel-io** [comment]
please 🙏 

its so annoying
